### PR TITLE
Create Faces 4.1 / Pull in 4.1.0-RC1

### LINKED ISSUE #27503: Create initial Faces 4.1 projects and bundles

### PR DESCRIPTION
**Pull request by volosied**
fixes #27503

Re-enable skipped tests: https://github.com/OpenLiberty/open-liberty/issues/27628
- AnnotationLiteralsTest

### CODE REVIEW

File: dev/io.openliberty.faces.4.1.thirdparty/.project [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+			<arguments>
+			</arguments>
+		</buildCommand>
+	</buildSpec>
+	<natures>
+		<nature>org.eclipse.jdt.core.javanature</nature>
+		<nature>bndtools.core.bndnature</nature>
+	</natures>
+</projectDescription>
+
```
Comment (volosied) [inline dev/io.openliberty.faces.4.1.thirdparty/.project]: New line

File: dev/io.openliberty.faces.4.1.thirdparty/.settings/org.eclipse.jdt.core.prefs [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+org.eclipse.jdt.core.compiler.codegen.inlineJsrBytecode=enabled
+org.eclipse.jdt.core.compiler.codegen.targetPlatform=11
+org.eclipse.jdt.core.compiler.codegen.unusedLocal=preserve
+org.eclipse.jdt.core.compiler.compliance=11
+org.eclipse.jdt.core.compiler.debug.lineNumber=generate
+org.eclipse.jdt.core.compiler.debug.localVariable=generate
+org.eclipse.jdt.core.compiler.debug.sourceFile=generate
+org.eclipse.jdt.core.compiler.problem.assertIdentifier=error
+org.eclipse.jdt.core.compiler.problem.enumIdentifier=error
+org.eclipse.jdt.core.compiler.source=11
```
Comment (volosied) [inline dev/io.openliberty.faces.4.1.thirdparty/.settings/org.eclipse.jdt.core.prefs]: 17

File: dev/io.openliberty.faces.4.1.thirdparty/bnd.bnd [hunk 3]
```diff
@@ -0,0 +1,45 @@
+#*******************************************************************************
+# Copyright (c) 2022, 2023 IBM Corporation and others.
```
Comment (volosied) [inline dev/io.openliberty.faces.4.1.thirdparty/bnd.bnd]: copyright

File: dev/io.openliberty.org.apache.myfaces.4.1/build.gradle [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+// Local vars
+File downloadSrcDir = project.file('build/source')
+
+configurations {
+    sourceJar
+}
+
+dependencies {
+   sourceJar 'org.apache.myfaces.core:myfaces-impl:4.1.0-RC1:sources'
```
Comment (volosied) [inline dev/io.openliberty.org.apache.myfaces.4.1/build.gradle:22]: comment out for now

**Comment by volosied** [comment]
bundles all resolve now, but I'm getting: 
```
[1/31/24, 16:13:36:941 EST] 00000049 com.ibm.ws.logging.internal.impl.IncidentImpl                I FFDC1015I: An FFDC Incident has been created: "java.lang.IllegalStateException: No Factories configured for this Application. This happens if the faces-initialization does not work at all - make sure that you properly include all configuration settings necessary for a basic faces application and that all the necessary libs are included. Also check the logging output of your web application and your container for any exceptions!
If you did that and find nothing, the mistake might be due to the fact that you use some special web-containers which do not support registering context-listeners via TLD files and a context listener is not setup in your web.xml.
A typical config looks like this;
<listener>
  <listener-class>org.apache.myfaces.webapp.StartupServletContextListener</listener-class>
</listener>
```

**Comment by volosied** [comment]
Most JSF 2.2 tests are pass now (have not looked at 2.3 yet), but I'm getting a few errors.   For reference later - here are the apps: 
[JSF22-4.1-Servers.zip](https://github.com/OpenLiberty/open-liberty/files/14172571/JSF22-4.1-Servers.zip)

First Exception is:

java.lang.RuntimeException: Non-success result code from Transformer: Transform Error
- Looks to be related to the new exception thrown here: 
- https://github.com/OpenLiberty/open-liberty/blob/integration/dev/fattest.simplicity/src/componenttest/rules/repeater/TransformSubActionImpl.java#L21-L24
- Need to debug to understand why the transformer failing? 

Once I revert that change, I encounter the following failures: 

http://localhost:9080/JSF22Miscellaneous/testViewScopeMyFaces.jsf 
- jakarta.el.PropertyNotFoundException: Target Unreachable, identifier [emailBean] resolved to null

http://localhost:9080/JSF22Miscellaneous/testAPI.xhtml
- jakarta.el.ELException: java.lang.NoClassDefFoundError: javax/faces/event/FacesListener 

http://localhost:9080/JSF22Miscellaneous/testCustomValueBindingTag.jsf
- java.lang.NoSuchMethodError - 'jakarta.faces.el.ValueBinding jakarta.faces.application.Application.createValueBinding(java.lang.String)'

http://localhost:9080/TestJSFEL/ComponentEventListener.xhtml 
- Caused by: java.lang.NoSuchMethodError: 'void com.ibm.ws.jsf22.el.components.CustomComponent$1.<init>(com.ibm.ws.jsf22.el.components.CustomComponent, jakarta.el.ExpressionFactory)'

File: dev/com.ibm.ws.jsf.2.2_fat/fat/src/com/ibm/ws/jsf22/fat/tests/JSFCompELTests.java [hunk 5]
```diff
@@ -141,6 +141,7 @@ protected void verifyResponse(String contextRoot, String resource, String... exp
         }
     }
 
+    //  VS FAIL The page did not contain the following expected response: The order and number of ELResolvers are correct!
```
Comment (volosied) [inline dev/com.ibm.ws.jsf.2.2_fat/fat/src/com/ibm/ws/jsf22/fat/tests/JSFCompELTests.java]: Investigate in personal build

File: dev/io.openliberty.faces.4.1.internal/src/io/openliberty/faces41/internal/cdi/WASCDIAnnotationInjectionProvider.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
+/*******************************************************************************
+ * Copyright (c) 2017, 2022 IBM Corporation and others.
```
Comment (volosied) [inline dev/io.openliberty.faces.4.1.internal/src/io/openliberty/faces41/internal/cdi/WASCDIAnnotationInjectionProvider.java]: These project files were copied from `io.openliberty.faces.4.0.internal`

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsf.2.2_fat,com.ibm.ws.jsf.2.3_fat

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsf.2.2_fat,com.ibm.ws.jsf.2.3_fat

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsf.2.2_fat,com.ibm.ws.jsf.2.3_fat,io.openliberty.org.apache.myfaces.4.0_fat

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsf.2.2_fat,com.ibm.ws.jsf.2.3_fat,io.openliberty.org.apache.myfaces.4.0_fat

**Comment by fritze2** [comment]
Kicking off another time because the previous runs failed on an infrastructure defect, I believe.
#build #spawn.fullfat.buckets=com.ibm.ws.jsf.2.2_fat,com.ibm.ws.jsf.2.3_fat,io.openliberty.org.apache.myfaces.4.0_fat

File: dev/io.openliberty.org.apache.myfaces.4.0_fat/bnd.bnd [hunk 7]
```diff
@@ -29,6 +29,14 @@ src: \
     test-applications/WebSocket.war/src,\
     test-applications/MYFACES-4628.war/src
 
+tested.features:\
+    expressionlanguage-6.0,\
+    faces-4.1,\
+    noship-1.0,\
```
Comment (volosied) [inline dev/io.openliberty.org.apache.myfaces.4.0_fat/bnd.bnd:35]: Remove noShip-1.0 from Faces FATs once we beta.

Comment (volosied) [reply dev/io.openliberty.org.apache.myfaces.4.0_fat/bnd.bnd:35]: https://github.com/OpenLiberty/open-liberty/issues/27639

File: dev/com.ibm.websphere.appserver.features/visibility/private/io.openliberty.jakarta.faces-4.1.feature [hunk 8]
```diff
@@ -3,7 +3,7 @@ symbolicName=io.openliberty.jakarta.faces-4.1
 singleton=true
 -features=com.ibm.websphere.appserver.eeCompatible-11.0, \
   io.openliberty.noShip-1.0
--bundles=io.openliberty.jakarta.faces.4.0; location:="dev/api/spec/,lib/"; mavenCoordinates="org.apache.myfaces.core:myfaces-api:4.0.0"
+-bundles=io.openliberty.jakarta.faces.4.1; location:="dev/api/spec/,lib/"; mavenCoordinates="org.apache.myfaces.core:myfaces-api:4.0.0"
```
Comment (pnicolucci) [inline dev/com.ibm.websphere.appserver.features/visibility/private/io.openliberty.jakarta.faces-4.1.feature]: mavenCoordinates need to be updated.

File: dev/com.ibm.ws.jsf.2.2_fat/fat/src/com/ibm/ws/jsf22/fat/tests/JSFServerTest.java [hunk 9]
```diff
@@ -12,6 +12,9 @@
 import static componenttest.annotation.SkipForRepeat.EE10_FEATURES;
 import static componenttest.annotation.SkipForRepeat.EE8_FEATURES;
 import static componenttest.annotation.SkipForRepeat.EE9_FEATURES;
+import static componenttest.annotation.SkipForRepeat.EE8_OR_LATER_FEATURES;
+import static componenttest.annotation.SkipForRepeat.EE9_OR_LATER_FEATURES;
```
Comment (pnicolucci) [inline dev/com.ibm.ws.jsf.2.2_fat/fat/src/com/ibm/ws/jsf22/fat/tests/JSFServerTest.java]: Looks like the EE9_OR_LATER_FEATURES can be removed as it is not used in this class.

File: dev/io.openliberty.jakarta.faces.4.1/build.gradle [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+File downloadSrcDir = project.file('build/source')
+File downloadJavadocDir = project.file('build/javadoc')
+
+configurations {
+    sourceJar
+    javadocJar
+}
+
+dependencies {
+//    sourceJar 'org.apache.myfaces.core:myfaces-api:4.1.0-RC1:sources'
```
Comment (pnicolucci) [inline dev/io.openliberty.jakarta.faces.4.1/build.gradle]: Let's remove the comment I see these in artifactory now.

File: dev/io.openliberty.org.apache.myfaces.4.0_fat/fat/src/io/openliberty/org/apache/myfaces40/fat/FATSuite.java [hunk 11]
```diff
@@ -74,7 +76,8 @@
 public class FATSuite {
 
     @ClassRule
-    public static RepeatTests repeat = RepeatTests.withoutModification();
+    public static RepeatTests repeat = RepeatTests.with(new EmptyAction().fullFATOnly())
```
Comment (pnicolucci) [inline dev/io.openliberty.org.apache.myfaces.4.0_fat/fat/src/io/openliberty/org/apache/myfaces40/fat/FATSuite.java]: If we run lite mode with Java11, then we'll fail here. 

`    public static RepeatTests repeat = RepeatTests.with(new EmptyAction().conditionalFullFATOnly(EmptyAction.GREATER_THAN_OR_EQUAL_JAVA_17))
                    .andWith(FeatureReplacementAction.EE11_FEATURES());`

File: dev/io.openliberty.org.apache.myfaces.4.0_fat/fat/src/io/openliberty/org/apache/myfaces40/fat/tests/AnnotationLiteralsTest.java [hunk 12]
```diff
@@ -37,6 +38,7 @@
  * Tests to ensure new Annotation Literals can be used in an expected way.
  */
 @RunWith(FATRunner.class)
+@SkipForRepeat(SkipForRepeat.EE11_FEATURES) // Errors in EE11
```
Comment (pnicolucci) [inline dev/io.openliberty.org.apache.myfaces.4.0_fat/fat/src/io/openliberty/org/apache/myfaces40/fat/tests/AnnotationLiteralsTest.java:41]: Do we have an issue opened to go back and address this?

File: dev/io.openliberty.org.apache.myfaces.4.1/bnd.bnd [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+    implementation:=com.ibm.ws.jsf.shared.ext.CdiExtension; \
+    provide:='com.ibm.ws.cdi.extension.WebSphereCDIExtension'; \
+    properties:="service.vendor=IBM,\
+    bean.defining.annotations=jakarta.faces.flow.FlowScoped;jakarta.faces.view.ViewScoped,\
+    api.classes=jakarta.faces.flow.builder.FlowDefinition;jakarta.faces.flow.builder.FlowBuilderParameter,\
+    extension.classes.only=true",\
+  io.openliberty.faces_40.dd; \
+    implementation:=com.ibm.ws.javaee.version.FacesVersion; \
+    provide:=com.ibm.ws.javaee.version.FacesVersion; \
+    properties:="version:Integer=40"
```
Comment (pnicolucci) [inline dev/io.openliberty.org.apache.myfaces.4.1/bnd.bnd]: Do we need updates here for 41? Can we at least open an issue for investigation?

File: dev/io.openliberty.org.apache.myfaces.4.1/bnd.bnd [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+
+# Use this format (4.1.0.RC#)  once we have an official release and not an RC Version
+Include-Resource: \
+  @${repo;org.apache.myfaces.core:myfaces-impl;4.1.0.RC1;EXACT}!/META-INF/**, \
+  @${repo;io.openliberty.jakarta.faces.4.0}!/META-INF/resources/**, \
+  @${repo;io.openliberty.jakarta.faces.4.0}!/META-INF/internal-resources/**, \
+  META-INF=@src/META-INF
+
+Service-Component: \
+  io.openliberty.faces_4_0.cdiextension; \
```
Comment (pnicolucci) [inline dev/io.openliberty.org.apache.myfaces.4.1/bnd.bnd]: Issue opened to look at updating this for 4.1?

File: dev/io.openliberty.org.apache.myfaces.4.1/bnd.bnd [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+  io.openliberty.faces40.internal.cdi, \
+  io.openliberty.faces40.internal.config.annotion, \
+  io.openliberty.faces40.internal.config.resource, \
+  io.openliberty.faces40.internal.ee, \
+  io.openliberty.faces40.internal.extprocessor, \
+  io.openliberty.faces40.internal.spi.impl
+
+# Don't export the org.apache.myfaces.buildtools package as that is only needed for compilation.  The jar that contains the
+# org.apache.myfaces.buildtools package is myfaces-builder-annotations-1.0.9.jar
+# Additionally, a subset of these org.apache.myfaces packages are exported by the thirdparty.faces-3.0 bundle - so we exclude them here.
```
Comment (pnicolucci) [inline dev/io.openliberty.org.apache.myfaces.4.1/bnd.bnd]: Update to references the thirdparty.faces-4.1 bundle?

File: dev/io.openliberty.org.apache.myfaces.4.1/bnd.bnd [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+  org.apache.myfaces.core:myfaces-impl;strategy=exact;version=4.1.0.RC1,\
+  io.openliberty.jakarta.faces.4.1;version=latest,\
+  com.ibm.ws.classloading;version=latest,\
+  com.ibm.ws.webcontainer.jakarta;version=latest,\
+  com.ibm.ws.serialization;version=latest,\
+  com.ibm.ws.container.service;version=latest,\
+  com.ibm.ws.adaptable.module;version=latest,\
+  com.ibm.ws.anno;version=latest,\
+  com.ibm.ws.managedobject;version=latest,\
+  io.openliberty.jakarta.expressionLanguage.5.0;version=latest,\
```
Comment (pnicolucci) [inline dev/io.openliberty.org.apache.myfaces.4.1/bnd.bnd]: expressionLanguage.6.0?

File: dev/io.openliberty.org.apache.myfaces.4.1/build.gradle [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+
+// Local vars
+File downloadSrcDir = project.file('build/source')
+
+configurations {
+    sourceJar
+}
+
+dependencies {
+//    sourceJar 'org.apache.myfaces.core:myfaces-impl:4.1.0-RC1:sources'
```
Comment (pnicolucci) [inline dev/io.openliberty.org.apache.myfaces.4.1/build.gradle]: uncomment if jars are available

File: dev/io.openliberty.org.apache.myfaces.4.1/src/META-INF/beans.xml [hunk 18]
```diff
@@ -0,0 +1,21 @@
+<!--
+    Copyright (c) 2022 IBM Corporation and others.
```
Comment (pnicolucci) [inline dev/io.openliberty.org.apache.myfaces.4.1/src/META-INF/beans.xml:2]: Is this just copied and the reason for no new copyright header?

File: dev/io.openliberty.org.apache.myfaces.4.1/src/META-INF/beans.xml [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+    which accompanies this distribution, and is available at
+    http://www.eclipse.org/legal/epl-2.0/
+    
+    SPDX-License-Identifier: EPL-2.0
+   
+    Contributors:
+        IBM Corporation - initial API and implementation
+ -->
+ <beans xmlns="https://jakarta.ee/xml/ns/jakartaee" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+ xsi:schemaLocation="https://jakarta.ee/xml/ns/jakartaee https://jakarta.ee/xml/ns/jakartaee/beans_4_0.xsd" version="4.0"
```
Comment (pnicolucci) [inline dev/io.openliberty.org.apache.myfaces.4.1/src/META-INF/beans.xml:14]: Do we need to update the version here? Or at least open an issue for investigation?

File: dev/io.openliberty.org.apache.myfaces.4.0_fat/fat/src/io/openliberty/org/apache/myfaces40/fat/tests/AnnotationLiteralsTest.java (hunk 12 above)
Comment (volosied) [reply dev/io.openliberty.org.apache.myfaces.4.0_fat/fat/src/io/openliberty/org/apache/myfaces40/fat/tests/AnnotationLiteralsTest.java:41]: Yes - https://github.com/OpenLiberty/open-liberty/issues/27628

File: dev/io.openliberty.org.apache.myfaces.4.1/src/META-INF/beans.xml (hunk 18 above)
Comment (volosied) [reply dev/io.openliberty.org.apache.myfaces.4.1/src/META-INF/beans.xml:2]: Correct - same beans.xml as with the other impl bundles.

File: dev/io.openliberty.org.apache.myfaces.4.1/src/META-INF/beans.xml (hunk 19 above)
Comment (volosied) [reply dev/io.openliberty.org.apache.myfaces.4.1/src/META-INF/beans.xml:14]: https://github.com/OpenLiberty/open-liberty/issues/27642

File: dev/io.openliberty.org.apache.myfaces.4.1/build.gradle (hunk 17 above)
Comment (volosied) [reply dev/io.openliberty.org.apache.myfaces.4.1/build.gradle]: Not yet, I've messaged Chuck again.

File: dev/io.openliberty.jakarta.faces.4.1/build.gradle (hunk 10 above)
Comment (volosied) [reply dev/io.openliberty.jakarta.faces.4.1/build.gradle]: I don't see them?  -- Edit: Now uploaded :)

**Review by pnicolucci — DISMISSED** [review]

File: dev/com.ibm.ws.jsf.2.3_fat/fat/src/com/ibm/ws/jsf23/fat/FATSuite.java [hunk 20]
```diff
@@ -105,23 +105,21 @@ public class FATSuite extends TestContainerSuite {
...
+            repeat = RepeatTests.with(new EmptyAction().fullFATOnly())
+                            .andWith(FeatureReplacementAction.EE10_FEATURES().conditionalFullFATOnly(FeatureReplacementAction.GREATER_THAN_OR_EQUAL_JAVA_17))
+                            .andWith(FeatureReplacementAction.EE11_FEATURES());
         } else {
+            // If we are running with a Java version less than 11 have EE9 be the lite mode test to run.
             repeat = RepeatTests.with(new EmptyAction().fullFATOnly())
-                            .andWith(FeatureReplacementAction.EE9_FEATURES());
+                            .andWith(FeatureReplacementAction.EE9_FEATURES().conditionalFullFATOnly(FeatureReplacementAction.GREATER_THAN_OR_EQUAL_JAVA_11))
+                            .andWith(FeatureReplacementAction.EE10_FEATURES().conditionalFullFATOnly(FeatureReplacementAction.GREATER_THAN_OR_EQUAL_JAVA_17))
+                            .andWith(FeatureReplacementAction.EE11_FEATURES());
```
Comment (pnicolucci) [inline dev/com.ibm.ws.jsf.2.3_fat/fat/src/com/ibm/ws/jsf23/fat/FATSuite.java:122]: If Java < 11 is used there is not lite mode defined when we skip EE9.

Comment (pnicolucci) [reply dev/com.ibm.ws.jsf.2.3_fat/fat/src/com/ibm/ws/jsf23/fat/FATSuite.java:122]:
```
// If we are running with a Java version less than 11 have EE8 (EmptyAction) be the lite mode test to run.
            repeat = RepeatTests.with(new EmptyAction().conditionalFullFATOnly(EmptyAction.GREATER_THAN_OR_EQUAL_JAVA_11))
                            .andWith(FeatureReplacementAction.EE10_FEATURES().conditionalFullFATOnly(FeatureReplacementAction.GREATER_THAN_OR_EQUAL_JAVA_17))
                            .andWith(FeatureReplacementAction.EE11_FEATURES());
```

**Review by pnicolucci — APPROVED** [review]

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsf.2.2_fat,com.ibm.ws.jsf.2.3_fat,io.openliberty.org.apache.myfaces.4.0_fat

**Comment by volosied** [comment]
#libby

**Pull request merged**